hoot-core/src/main/cpp/hoot/core/io/OsmApiDb.cpp

```diff
@@ -617,4 +626,144 @@ double OsmApiDb::fromOsmApiDbCoord(const long x)
   return (double)x / COORDINATE_SCALE;
 }
 
+shared_ptr<QSqlQuery> OsmApiDb::selectNodesByBounds(const QString bbox)
+{
+  const Envelope env = GeometryUtils::envelopeFromConfigString(bbox);
+  LOG_VARD(env);
+  const vector<Range> tileRanges = _getTileRanges(env);
+  LOG_VARD(tileRanges.size());
+
+  //I'm not sure yet if the number of tile ID ranges is constant or not.  If it is, then we could
+  //modify this to add placeholders for the tile ID ranges and make it so the query only gets
+  //prepared once.
+  if (!_selectNodesByBounds)
+  {
+    _selectNodesByBounds.reset(new QSqlQuery(_db));
+    _selectNodesByBounds->setForwardOnly(true);
+  }
+  QString sql = "select * from current_nodes where";
+  sql += " visible = true";
+  sql += " and (" + _getTileWhereCondition(tileRanges) + ")";
+  sql += " and longitude >= :minLon and longitude <= :maxLon and latitude >= :minLat ";
+  sql += " and latitude <= :maxLat";
+  sql += " order by id desc";
+  _selectNodesByBounds->prepare(sql);
+  _selectNodesByBounds->bindValue(":minLon", (qlonglong)(env.getMinX() * COORDINATE_SCALE));
+  _selectNodesByBounds->bindValue(":maxLon", (qlonglong)(env.getMaxX() * COORDINATE_SCALE));
+  _selectNodesByBounds->bindValue(":minLat", (qlonglong)(env.getMinY() * COORDINATE_SCALE));
+  _selectNodesByBounds->bindValue(":maxLat", (qlonglong)(env.getMaxY() * COORDINATE_SCALE));
+  LOG_VARD(_selectNodesByBounds->lastQuery());
+  LOG_VARD(_selectNodesByBounds->boundValues());
+
+  if (_selectNodesByBounds->exec() == false)
+  {
+    throw HootException(
+      "Error selecting nodes by bounds: " + bbox + " Error: " +
+      _selectNodesByBounds->lastError().text());
+  }
+  LOG_VARD(_selectNodesByBounds->numRowsAffected());
+  return _selectNodesByBounds;
+}
+
+shared_ptr<QSqlQuery> OsmApiDb::selectWayIdsByWayNodeIds(const QStringList& nodeIds)
+{
+  if (!_selectWayIdsByWayNodeIds)
+  {
+    _selectWayIdsByWayNodeIds.reset(new QSqlQuery(_db));
+    _selectWayIdsByWayNodeIds->setForwardOnly(true);
+  }
+  //this has to be prepared every time due to the varying number of IDs passed in
+  QString sql =
+    "select way_id from current_way_nodes where node_id in (" + nodeIds.join(",") + ")";
```

Comment:
I suspect that this SQL statement should have the ```DISTINCT``` keyword in it - as in ```SELECT DISTINCT way_id from ...```

Because you could have multiple nodes belonging to one way in your nodeIds list, which would result in the same way_id being returned several times.